cmapPy/math/fast_corr.py

```diff
@@ -0,0 +1,43 @@
+import logging
+import cmapPy.pandasGEXpress.setup_GCToo_logger as setup_logger
+import numpy
+import cmapPy.math.fast_cov as fast_cov
+
+
+logger = logging.getLogger(setup_logger.LOGGER_NAME)
+
+
+def fast_corr(x, y=None, axis=0, do_print_trace=False):
+    """calculate the correlation matrix for x, or optionally, the correlaton matrix between x and y.
+    axis parameter defines whether to treat the rows or columns as vectors that the covariance is calculated
+    between
+
+    Args:
+        x (numpy array-like) NxM in shape
+        y (numpy array-like) OxP in shape
+        axis (integer) indicates which axis of the arrays is the vectors.  axis=0 indicates calculate
+            the covariance between the rows, axis=1 indicates calculate the covariance between the columns
+        do_print_trace (bool) whether or not to use logger.debug to print out intermediate mathematical steps
+            (WARNING: will print out all intermediate matrices, potentially lots of output!)
+
+        returns (numpy array-like) QxR array of the covariance values
+            for defaults (y=None, axis=0), shape is NxN
+            if y=None and axis=1, shape is MxM
+            if y is provied, axis=0, shape is NxO (and P==M in other words number of columns in x & y are the same)
+            if y is provided, axis=1, shape is MxP (and N==O in other words number of rows in x & y are the same)
+    """
+    if y is None:
+        y = x
+
+    cov_mat = fast_cov.fast_cov(x, y, axis, do_print_trace)
+    if do_print_trace:  logger.debug("cov_mat:\n{}".format(cov_mat))
+    
+    std_x = numpy.std(x.T, axis=axis, ddof=1)
+    if do_print_trace:  logger.debug("std_x:\n{}".format(std_x))
+    std_y = numpy.std(y.T, axis=axis, ddof=1)
+    if do_print_trace:  logger.debug("std_y:\n{}".format(std_y))
+    
+    std_outer = numpy.outer(std_x, std_y)
+    if do_print_trace:  logger.debug("std_outer:\n{}".format(std_outer))
+
+    return cov_mat / std_outer
```

Comment:
Not sure where (maybe as an argument? Not sure how that would work with the fast aspect of it) but I think it's important to mention this is pearson correlation. I think documentation should include it at the very least, thoughts?

Reply:
I'll definitely add to the doc string that this is doing pearson.  I wasn't sure exactly the best way to add support for spearman - I was thinking as a separate module / method that calculates the ranks, mean centers them, and then calls fast_corr.

Reply:
good call. adding spearman would be CLUTCH!

Reply:
re: spearman, I would use scipy.stats.rankdata, but that means we need to add scipy to the cmapPy environment.  I don't think that's a big deal but wanted you all to consider it.
We could do some (slight) gymnastics with numpy.argsort to get there as well, I think.  Maybe.

Reply:
As long as it doesn't break anything I don't see adding scipy to the environment as a problem

Reply:
So I ended up not using scipy to calculate the ranks - I had forgotten that it's only option is to flatten the array first, I used pandas to do the calculation.